File: src/main/java/alfheim/api/spell/DamageSourceSpell.java
package alfheim.api.spell;

import alfheim.common.entity.spell.EntitySpellDriftingMine;
import alfheim.common.entity.spell.EntitySpellFenrirStorm;
import alfheim.common.entity.spell.EntitySpellFireball;
import alfheim.common.entity.spell.EntitySpellFirewall;
import alfheim.common.entity.spell.EntitySpellMortar;
import alfheim.common.entity.spell.EntitySpellWindBlade;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ChatComponentTranslation;
import net.minecraft.util.DamageSource;
import net.minecraft.util.EntityDamageSource;
import net.minecraft.util.EntityDamageSourceIndirect;
import net.minecraft.util.IChatComponent;
import net.minecraft.util.StatCollector;
import net.minecraft.world.Explosion;

public class DamageSourceSpell extends DamageSource {

	public DamageSourceSpell(String type) {
		super(type);
	}

	public static final DamageSource bleeding	= new DamageSource		("bleeding")	.setDamageBypassesArmor().setDamageIsAbsolute();
	public static final DamageSource gravity	= new DamageSourceSpell	("gravity")		.setDamageBypassesArmor().setDifficultyScaled();
	public static final DamageSource mark		= new DamageSourceSpell	("mark")		.setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage();
	public static final DamageSource poison		= new DamageSourceSpell	("poison")		.setDamageBypassesArmor();
	public static final DamageSource possession = new DamageSourceSpell	("possession")	.setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage();
	public static final DamageSource sacrifice	= new DamageSourceSpell	("sacrifice")	.setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage();
	public static final DamageSource soulburn	= new DamageSourceSpell	("soulburn")	.setDamageBypassesArmor().setMagicDamage();

	public static DamageSource blades(EntitySpellWindBlade wb, EntityLivingBase caster) {
		return new EntityDamageSourceIndirectSpell("windblade", caster, wb).setDamageBypassesArmor();
	}
	
	/** Sacrifice type of damage to attack other mobs */
	public static DamageSource darkness(EntityLivingBase attacker) {
		return new EntityDamageSourceSpell("darkness_FF", attacker).setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage();
	}
	
	public static DamageSource explosion(EntitySpellDriftingMine dm, EntityLivingBase caster) {
        return new EntityDamageSource("explosion.player", caster).setDifficultyScaled().setExplosion();
    }
	
    public static DamageSource fireball(EntitySpellFireball fb, EntityLivingBase caster) {
        return new EntityDamageSourceIndirectSpell("fireball", caster, fb).setFireDamage().setExplosion().setProjectile();
    }
	
    public static DamageSource firewall(EntitySpellFirewall fw, EntityLivingBase caster) {
        return new EntityDamageSourceIndirectSpell("firewall", caster, fw).setFireDamage();
    }
    
	/** Fenrir Storm type of damage */
	public static DamageSource lightning(EntitySpellFenrirStorm st, EntityLivingBase caster) {
		return new EntityDamageSourceIndirectSpell("lightning", caster, st).setDamageBypassesArmor().setFireDamage();
	}
	
	public static DamageSource mortar(EntitySpellMortar mt, EntityLivingBase caster) {
		return new EntityDamageSourceIndirectSpell("fallingBlock", caster, mt).setDifficultyScaled().setProjectile();
	}
	
	/** Some water blades (?) type of damage */
	public static DamageSource water(EntityLivingBase caster) {
		return (new EntityDamageSourceSpell("water", caster)).setDamageBypassesArmor();
	}
	
	public static class EntityDamageSourceSpell extends DamageSourceSpell {
	    protected Entity attacker;

	    public EntityDamageSourceSpell(String source, Entity attacker) {
	        super(source);
	        this.attacker = attacker;
	    }

	    public Entity getEntity() {
	        return this.attacker;
	    }

	    public IChatComponent func_151519_b(EntityLivingBase target) {
	        ItemStack itemstack = this.attacker instanceof EntityLivingBase ? ((EntityLivingBase)this.attacker).getHeldItem() : null;
	        String s = "death.attack." + this.damageType;
	        String s1 = s + ".item";
	        return itemstack != null && itemstack.hasDisplayName() && StatCollector.canTranslate(s1) ? new ChatComponentTranslation(s1, new Object[] {target.func_145748_c_(), this.attacker.func_145748_c_(), itemstack.func_151000_E()}): new ChatComponentTranslation(s, new Object[] {target.func_145748_c_(), this.attacker.func_145748_c_()});
	    }

	    public boolean isDifficultyScaled() {
	        return this.attacker != null && this.attacker instanceof EntityLivingBase && !(this.attacker instanceof EntityPlayer);
	    }
	}
	
	public static class EntityDamageSourceIndirectSpell extends EntityDamageSourceSpell {
	    private Entity indirectEntity;

	    public EntityDamageSourceIndirectSpell(String type, Entity attacker, Entity indirect) {
	        super(type, attacker);
	        this.indirectEntity = indirect;
	    }

	    public Entity getSourceOfDamage() {
	        return this.attacker;
	    }

	    public Entity getEntity() {
	        return this.indirectEntity;
	    }

	    public IChatComponent func_151519_b(EntityLivingBase target) {
	        IChatComponent ichatcomponent = this.indirectEntity == null ? this.attacker.func_145748_c_() : this.indirectEntity.func_145748_c_();
	        ItemStack itemstack = this.indirectEntity instanceof EntityLivingBase ? ((EntityLivingBase)this.indirectEntity).getHeldItem() : null;
	        String s = "death.attack." + this.damageType;
	        String s1 = s + ".item";
	        return itemstack != null && itemstack.hasDisplayName() && StatCollector.canTranslate(s1) ? new ChatComponentTranslation(s1, new Object[] {target.func_145748_c_(), ichatcomponent, itemstack.func_151000_E()}): new ChatComponentTranslation(s, new Object[] {target.func_145748_c_(), ichatcomponent});
	    }
	}
}

File: src/main/java/alfheim/common/core/util/DamageSourceSpell.kt
package alfheim.common.core.util

import alfheim.common.entity.spell.*
import net.minecraft.entity.*
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.*

open class DamageSourceSpell(type: String): DamageSource(type) {
	
	companion object {
		
		/** Any anomaly */
		val anomaly = DamageSource("anomaly").setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage()!!
		/** Decay Spell */
		val bleeding = DamageSourceSpell("bleeding").setDamageBypassesArmor().setDamageIsAbsolute()!!
		/** Death Mark Spell */
		val mark = DamageSourceSpell("mark").setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage()!!
		/** Regular poison */
		val poison = DamageSource("poison").setDamageBypassesArmor()!!
		/** Magical poison */
		val poisonMagic = DamageSourceSpell("poison").setDamageBypassesArmor()!!
		/** Tank Mask */
		val possession = DamageSource("possession").setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage()!!
		/** Sacrifice Spell */
		val sacrifice = DamageSourceSpell("sacrifice").setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage()!!
		/** Red Flame */
		val soulburn = DamageSource("soulburn").setDamageBypassesArmor().setMagicDamage()!!
		
		fun explosion(dm: EntitySpellDriftingMine, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("explosion.player", caster, dm).setExplosion()!!
		
		fun fireball(fb: EntitySpellFireball, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("fireball", caster, fb).setFireDamage().setExplosion().setProjectile()!!
		
		fun firewall(fw: EntitySpellFirewall, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("firewall", caster, fw).setFireDamage()!!
		
		fun gravity(gt: EntitySpellGravityTrap, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("gravity", caster, gt).setDamageBypassesArmor()!!
		
		fun hammerfall(caster: EntityLivingBase?) =
			EntityDamageSourceSpell("hammerfall", caster).setDamageBypassesArmor().setProjectile()!!
		
		/** Fenrir Storm type of damage  */
		fun lightning(st: EntitySpellFenrirStorm, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("lightning", caster, st).setDamageBypassesArmor().setFireDamage()!!
		
		fun missile(im: EntitySpellIsaacMissile, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("missile", caster, im).setMagicDamage()!!
		
		fun mortar(mt: EntitySpellMortar, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("mortar", caster, mt).setProjectile()!!
		
		fun sacrifice(caster: EntityLivingBase?) =
			EntityDamageSourceSpell("darkness_FF", caster).setDamageBypassesArmor().setDamageIsAbsolute().setMagicDamage()!!
		
		fun shadow(caster: EntityLivingBase?) =
			EntityDamageSource("shadow", caster).setDamageBypassesArmor().setMagicDamage()!!
		
		/** Shadow vortex type of damage */
		fun shadowSpell(caster: EntityLivingBase?) =
			EntityDamageSourceSpell("shadow", caster).setDamageBypassesArmor().setMagicDamage()!!
		
		/** Some water blades (?) type of damage  */
		fun water(caster: EntityLivingBase?) =
			EntityDamageSourceSpell("water", caster).setDamageBypassesArmor()!!
		
		fun windblade(wb: EntitySpellWindBlade, caster: EntityLivingBase?) =
			EntityDamageSourceIndirectSpell("windblade", caster, wb).setDamageBypassesArmor()!!
	}
}

open class EntityDamageSourceSpell(source: String, protected val attacker: Entity?): DamageSourceSpell(source) {
	
	override fun getEntity() = attacker
	
	override fun func_151519_b(target: EntityLivingBase): IChatComponent {
		val itemstack = if (attacker is EntityLivingBase) attacker.heldItem else null
		val s = "death.attack.$damageType"
		val s1 = "$s.item"
		return if (itemstack != null && itemstack.hasDisplayName() && StatCollector.canTranslate(s1)) ChatComponentTranslation(s1, target.func_145748_c_(), attacker!!.func_145748_c_(), itemstack.func_151000_E()) else ChatComponentTranslation(s, target.func_145748_c_(), attacker!!.func_145748_c_())
	}
	
	override fun isDifficultyScaled() =
		attacker != null && attacker is EntityLivingBase && attacker !is EntityPlayer
}

open class EntityDamageSourceIndirectSpell(type: String, attacker: Entity?, protected val dealer: Entity?): EntityDamageSourceSpell(type, attacker) {
	
	override fun getSourceOfDamage() = dealer
	
	override fun func_151519_b(target: EntityLivingBase): IChatComponent {
		val ichatcomponent = if (attacker == null) dealer!!.func_145748_c_() else attacker.func_145748_c_()
		val itemstack = if (attacker is EntityLivingBase) attacker.heldItem else null
		val s = "death.attack.$damageType"
		val s1 = "$s.item"
		return if (itemstack != null && itemstack.hasDisplayName() && StatCollector.canTranslate(s1)) ChatComponentTranslation(s1, target.func_145748_c_(), ichatcomponent, itemstack.func_151000_E()) else ChatComponentTranslation(s, target.func_145748_c_(), ichatcomponent)
	}
}